### PR TITLE
Media Section: v1

### LINKED ISSUE #8381: Media: create new media section
Let's stub out a new media section! This requires creating a feature flag and some basic section configuration. See the config under: `wordpress-com.js`

### PR DESCRIPTION
**Pull request by artpi**
Resolves #8381. Introduces media section in Sidebar and route:
http://calypso.localhost:3000/media/:site

Currently only in development & wpcalypso
- Picks up @aduth work
- Ties existing components from `post-editor` and `media-modal`
## Pics or it didn't happen

zrzut ekranu 2016-10-30 o 14 27 22
### [Video](https://cloudup.com/iq1Pt8XiAHa)

(lags are because of my crappy mexican wifi)
## Testing

Just try to use media section like you always wanted.
## Future work
- Icon (Need Gridicon, @amylaneio design has a dashicon "media", which we didnt port to gridicons yet)
- Add button
- Space nudge at the bottom

CC @adambecker @amylaneio @lamosty @gwwar @mtias @retrofox @rralian

### CODE REVIEW

File: client/my-sites/media/main.jsx [hunk 1]
```diff
@@ -1,17 +1,19 @@
...
+import SidebarNavigation from 'my-sites/sidebar-navigation';
+import observe from 'lib/mixins/data-observe';
+import Dialog from 'components/dialog';
+import MediaModalDetail from 'post-editor/media-modal/detail';
 
-module.exports = React.createClass( {
+export default React.createClass( {
 	displayName: 'Media',
 
 	mixins: [ observe( 'sites' ) ],
```
Comment (gwwar) [inline client/my-sites/media/main.jsx]: We can actually get rid of this, since this is only using selected site which is available from our Redux tree. Without the mixin, we can then port this over as a plain js class. 

Up to you if you want to leave these as TODO's for future PR's though.

File: client/my-sites/sidebar/publish-menu.jsx [hunk 2]
```diff
@@ -78,6 +78,18 @@ const PublishMenu = React.createClass( {
...
+				name: 'media',
+				label: this.translate( 'Media' ),
+				className: 'media',
+				capability: 'upload_files',
+				queryable: true,
+				config: 'manage/media',
+				link: '/media',
+				//buttonLink: site ? '/media/' + site.slug : '/media',
+				wpAdminLink: 'upload.php',
+				showOnAllMySites: true,
```
Comment (gwwar) [inline client/my-sites/sidebar/publish-menu.jsx]: @aduth any reason why this route was left enabled before?
```
	if ( config.isEnabled( 'manage/media' ) ) {
		page( '/media', controller.siteSelection, controller.sites );
		page( '/media/:filter?/:domain', controller.siteSelection, controller.navigation, mediaController.media );
	}
```

File: config/wpcalypso.json [hunk 3]
```diff
@@ -49,6 +49,7 @@
 		"manage/import/medium": true,
 		"manage/jetpack": true,
 		"manage/jetpack-plans": true,
+		"manage/media": true,
```
Comment (gwwar) [inline config/wpcalypso.json]: Let's also turn this off for wpcalypso/dev, since we'll be focusing on other things for a while.

**Comment by artpi** [comment]
Design to keep in mind
Sourced from p5fxDp-rL-p2 by @amylaneio 

media-calypso

File: client/my-sites/sidebar/publish-menu.jsx (hunk 2 above)
Comment (artpi) [inline client/my-sites/sidebar/publish-menu.jsx]: This flag is *not* combined with the editor media modal. It was already here, referring only to this.
There is no other functionality using this flag AFAIK - please do a grep check to make sure.

File: client/my-sites/media-library/style.scss [hunk 4]
```diff
@@ -5,6 +5,10 @@
 @import 'upload-button';
 @import 'upload-url';
 
+.main.media {
```
Comment (aduth) [inline client/my-sites/media-library/style.scss]: This should probably be in `client/my-sites/media/style.scss`, since it's not specific to the modal.

File: client/my-sites/media/main.jsx [hunk 5]
```diff
@@ -44,14 +52,91 @@ module.exports = React.createClass( {
...
+							onDone={ this.onImageEditorDone }
+							onCancel={ this.onImageEditorCancel }
+						/>
+					}
+					</Dialog>
+				}
 				<MediaLibrary
 					{ ...this.props }
 					onFilterChange={ this.onFilterChange }
-					site={ this.props.sites.getSelectedSite() || undefined }
```
Comment (aduth) [inline client/my-sites/media/main.jsx]: The fallback to `undefined` is because `SitesList#getSelectedSite` can return `false`, thereby causing a warning to be logged by unexpected prop type of `<MediaLibrary />`.

File: client/my-sites/media/main.jsx [hunk 6]
```diff
@@ -44,14 +52,91 @@ module.exports = React.createClass( {
...
+
+	closeDetailsModal() {
+		this.setState( { openedDetails: null, editedItem: null } );
+	},
+
+	editImage() {
+		this.setState( { openedDetails: null, editedItem: this.state.openedDetails } );
+	},
+
+	onImageEditorCancel: function( imageEditorProps ) {
```
Comment (aduth) [inline client/my-sites/media/main.jsx]: I find it strange that the image editor's own props are being passed up back through the callback. Presumably the parent component rendering the image editor knows those props already or can determine them itself via `connect`?

File: client/my-sites/sidebar/publish-menu.jsx [hunk 7]
```diff
@@ -78,6 +78,18 @@ const PublishMenu = React.createClass( {
...
+			},
+			{
+				name: 'media',
+				label: this.translate( 'Media' ),
+				className: 'media',
+				capability: 'upload_files',
+				queryable: true,
+				config: 'manage/media',
+				link: '/media',
+				//buttonLink: site ? '/media/' + site.slug : '/media',
```
Comment (aduth) [inline client/my-sites/sidebar/publish-menu.jsx]: Remove commented line?

File: client/my-sites/sidebar/publish-menu.jsx (hunk 2 above)
Comment (aduth) [inline client/my-sites/sidebar/publish-menu.jsx]: My mistake, you're right. This config flag was meant for the standalone page.

Unrelated comment for this line: Since the page doesn't support an all sites view, I don't think it should be included in the all sites publish bar.

File: client/my-sites/sidebar/publish-menu.jsx [hunk 8]
```diff
@@ -78,6 +78,18 @@ const PublishMenu = React.createClass( {
 				buttonLink: site ? '/page/' + site.slug : '/page',
 				wpAdminLink: 'edit.php?post_type=page',
 				showOnAllMySites: true,
+			},
+			{
+				name: 'media',
+				label: this.translate( 'Media' ),
+				className: 'media',
```
Comment (aduth) [inline client/my-sites/sidebar/publish-menu.jsx]: I suppose it's consistent with other sidebar elements, but these simple class names seem ripe for conflict and don't really serve much purpose in their usage here.

File: client/post-editor/media-modal/detail/index.jsx [hunk 9]
```diff
@@ -16,7 +16,7 @@ var DetailItem = require( './detail-item' ),
 import { ModalViews } from 'state/ui/media-modal/constants';
 import { setEditorMediaModalView } from 'state/ui/editor/actions';
 
-const EditorMediaModalDetail = React.createClass( {
+export const EditorMediaModalDetail = React.createClass( {
```
Comment (aduth) [inline client/post-editor/media-modal/detail/index.jsx]: This breaks the detail view in the editor, since [the editor modal imports this file via CommonJS](https://github.com/Automattic/wp-calypso/blob/cf2a0b128e0525a5dbfdf241f5b594f34c02c325/client/post-editor/media-modal/index.jsx#L18). When exporting both default and named exports, we must either change all consuming modules to import via ES6 imports, or reference the `.default` or `.EditorMediaModalDetail` export.

Ref: p1477925766002698-slack-calypso-framework

File: client/my-sites/media/main.jsx (hunk 6 above)
Comment (artpi) [inline client/my-sites/media/main.jsx]: I agree with you, but that's an issue with ImageEditor design and I'd like not to take it on in this PR :)

File: client/my-sites/sidebar/publish-menu.jsx (hunk 7 above)
Comment (artpi) [inline client/my-sites/sidebar/publish-menu.jsx]: Yes, my bad. Will remove.

File: client/my-sites/sidebar/publish-menu.jsx (hunk 2 above)
Comment (artpi) [inline client/my-sites/sidebar/publish-menu.jsx]: Will change!

File: client/post-editor/media-modal/detail/index.jsx (hunk 9 above)
Comment (artpi) [inline client/post-editor/media-modal/detail/index.jsx]: Oh gosh! It didn't even occur for me to check!
Funny the tests passed though :/
I will definitely fix that!

File: client/my-sites/media/main.jsx (hunk 6 above)
Comment (gwwar) [inline client/my-sites/media/main.jsx]: I debated on that for a bit in https://github.com/Automattic/wp-calypso/pull/8689#pullrequestreview-4159039, but I'd be open to making the caller more knowledgeable about the editor, and keeping the callbacks clean provided we provide a few more examples of usage in the readme. I agree with @artpi though that this should be tackled in another PR.

File: client/my-sites/sidebar/publish-menu.jsx (hunk 8 above)
Comment (artpi) [inline client/my-sites/sidebar/publish-menu.jsx]: a85f987cf34bf8b821d6e4b55fe832955e56ecf9

File: client/post-editor/media-modal/detail/index.jsx (hunk 9 above)
Comment (artpi) [inline client/post-editor/media-modal/detail/index.jsx]: Done fa2a965

File: client/my-sites/sidebar/publish-menu.jsx (hunk 7 above)
Comment (artpi) [inline client/my-sites/sidebar/publish-menu.jsx]: Done 	d9d7d1b

File: client/my-sites/media-library/style.scss (hunk 4 above)
Comment (artpi) [inline client/my-sites/media-library/style.scss]: Done 0359523

**Comment by gwwar** [comment]
Pretty minor (can be deferred to another PR), but I think getting rid of the margin so the image previews lines up with the Media library title would be nice.
screen shot 2016-11-01 at 4 30 57 pm

Or maybe add some padding to the Media Library at this in-between width:
screen shot 2016-11-01 at 4 32 48 pm

File: config/wpcalypso.json [hunk 10]
```diff
@@ -49,6 +49,7 @@
 		"manage/export/guided-transfer": true,
 		"manage/import/medium": true,
 		"manage/jetpack": true,
+		"manage/media": true,
```
Comment (gwwar) [inline config/wpcalypso.json]: For anyone reading this would turn on the section for everything but Deskftop, Prod and Stage

**Comment by gwwar** [comment]
No clear way to edit at small sizes:
screen shot 2016-11-01 at 4 41 43 pm

**Comment by gwwar** [comment]
Not sure what to do with this one, but wanted to point out that the media is the only one that doesn't have an add button in the publish section:
screen shot 2016-11-01 at 4 40 04 pm

**Comment by gwwar** [comment]
👍 This chunk of work looks good to me. @aduth did you want to take another look? I haven't found any regressions so far.

**Comment by artpi** [comment]
> but wanted to point out that the media is the only one that doesn't have an add button in the publish section:

I will definitely open follow-up PR with that, but it will need some refactors

**Comment by gwwar** [comment]
I think it's fine to follow up on these in another PR. If we think the current implementation is still too rough we can change the envs this is available on to dev only, or off in all.

If you beat me to it, lets put issues we've noticed in the TODO column of https://github.com/Automattic/wp-calypso/projects/5

File: client/post-editor/media-modal/index.jsx [hunk 11]
```diff
@@ -15,7 +15,7 @@ var MediaLibrary = require( 'my-sites/media-library' ),
 	PostActions = require( 'lib/posts/actions' ),
 	PostStats = require( 'lib/posts/stats' ),
 	MediaModalSecondaryActions = require( './secondary-actions' ),
-	MediaModalDetail = require( './detail' ),
+	MediaModalDetail = require( './detail' ).default,
```
Comment (gwwar) [inline client/post-editor/media-modal/index.jsx]: Could we change this to an import below. This will also break if we only export one thing from MediaModalDetail

Comment (artpi) [inline client/post-editor/media-modal/index.jsx]: Done ! f85ab80
